xls/ir/block_test.cc

```diff
@@ -1002,5 +1002,99 @@ block my_block(in0: bits[32], in1: bits[32], out0: bits[32], out1: bits[32]) {
 )");
 }
 
+TEST_F(BlockTest, ReplaceInstantiationWithRename) {
+  auto p = CreatePackage();
+  Type* u32 = p->GetBitsType(32);
+
+  BlockBuilder sub_bb("sub_block", p.get());
+  {
+    BValue a = sub_bb.InputPort("a", u32);
+    BValue b = sub_bb.InputPort("b", u32);
+    sub_bb.OutputPort("x", a);
+    sub_bb.OutputPort("y", b);
+  }
+  XLS_ASSERT_OK_AND_ASSIGN(Block * sub_block, sub_bb.Build());
+  BlockBuilder add_bb("add_block", p.get());
+  {
+    BValue a_renamed = add_bb.InputPort("a_renamed", u32);
+    BValue b = add_bb.InputPort("b", u32);
+    add_bb.OutputPort("x_renamed", add_bb.Add(a_renamed, b));
+    add_bb.OutputPort("y", b);
+  }
+  XLS_ASSERT_OK_AND_ASSIGN(Block * add_block, add_bb.Build());
+
+  BlockBuilder bb("my_block", p.get());
+  XLS_ASSERT_OK_AND_ASSIGN(
+      Instantiation * instantiation,
+      bb.block()->AddBlockInstantiation("inst", sub_block));
+  {
+    BValue in0 = bb.InputPort("in0", u32);
+    BValue in1 = bb.InputPort("in1", u32);
+    bb.InstantiationInput(instantiation, "a", in0);
+    BValue out0 = bb.InstantiationOutput(instantiation, "x");
+    bb.InstantiationInput(instantiation, "b", in1);
+    BValue out1 = bb.InstantiationOutput(instantiation, "y");
+    bb.OutputPort("out0", out0);
+    bb.OutputPort("out1", out1);
+  }
+  XLS_ASSERT_OK_AND_ASSIGN(Block * block, bb.Build());
+
+  XLS_ASSERT_OK_AND_ASSIGN(auto inst_add,
+                           block->AddBlockInstantiation("inst_add", add_block));
+
+  // Replacing without rename should fail since port names do not match:
+  EXPECT_THAT(block->ReplaceInstantiationWith(instantiation, inst_add),
+              absl_testing::StatusIs(absl::StatusCode::kInternal,
+                                     testing::ContainsRegex("Type mismatch")));
+
+  // Replacing with rename of non-existent port should fail:
+  EXPECT_THAT(block->ReplaceInstantiationWith(instantiation, inst_add,
+                                              {{"does_not_exist", "a2"}}),
+              absl_testing::StatusIs(
+                  absl::StatusCode::kInternal,
+                  testing::ContainsRegex("rename port that does not exist")));
+
+  // Replacing with rename of port to used name should fail:
+  EXPECT_THAT(
+      block->ReplaceInstantiationWith(instantiation, inst_add, {{"a", "b"}}),
+      absl_testing::StatusIs(absl::StatusCode::kInternal,
+                             testing::ContainsRegex("name already exists")));
+
+  // Replacing with acceptable rename:
+  EXPECT_THAT(
+      block->ReplaceInstantiationWith(instantiation, inst_add,
+                                      {{"a", "a_renamed"}, {"x", "x_renamed"}}),
+      absl_testing::IsOk());
+  EXPECT_EQ(p->DumpIr(), R"(package ReplaceInstantiationWithRename
```

Comment:
Use golden-testfiles please.

See https://github.com/google/xls/blob/76f4486dbcaaa16705ca4fd2bd526aa24f2f45a9/xls/ir/node_util_test.cc#L295C1-L296C1 for example. This makes it easier to update if we change the way we format ir or something.